src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -0,0 +1,1054 @@
+      using ExpressionDataType = typename decltype(data_type_t)::type;
+
+      const auto expression_result = evaluate_expression_to_result<ExpressionDataType>(expression);
+      fn(*expression_result);
+    });
+  }
+}
+
+template <typename... RowCounts>
+ChunkOffset ExpressionEvaluator::_result_size(const RowCounts... row_counts) {
+  // If any operand is empty (that's the case IFF it is an empty column) the result of the expression has no rows
+  if (((row_counts == 0) || ...)) return 0;
+
+  return std::max({row_counts...});
+}
+
+std::vector<bool> ExpressionEvaluator::_evaluate_default_null_logic(const std::vector<bool>& left,
+                                                                    const std::vector<bool>& right) const {
+  const auto result_size = _result_size(left.size(), right.size());
+
+  if (result_size == 0) return {};
+
+  if (left.size() == right.size()) {
+    std::vector<bool> nulls(result_size);
+    std::transform(left.begin(), left.end(), right.begin(), nulls.begin(), [](auto l, auto r) { return l || r; });
+    return nulls;
+  } else if (left.size() > right.size()) {
+    DebugAssert(right.size() == 1,
+                "Operand should have either the same row count as the other or 1 (to represent a literal)");
+    if (right.front())
+      return std::vector<bool>({true});
+    else
+      return left;
+  } else {
+    DebugAssert(left.size() == 1,
+                "Operand should have either the same row count as the other or 1 (to represent a literal)");
+    if (left.front())
+      return std::vector<bool>({true});
+    else
+      return right;
+  }
+}
+
+void ExpressionEvaluator::_ensure_column_materialization(const ColumnID column_id) {
+  Assert(_chunk, "Cannot access Columns in this Expression as it doesn't operate on a Table/Chunk");
+
+  if (_column_materializations[column_id]) return;
+
+  const auto& column = *_chunk->get_column(column_id);
+
+  resolve_data_type(column.data_type(), [&](const auto column_data_type_t) {
+    using ColumnDataType = typename decltype(column_data_type_t)::type;
+
+    std::vector<ColumnDataType> values;
+    materialize_values(column, values);
+
+    if (_table->column_is_nullable(column_id)) {
+      std::vector<bool> nulls;
+      materialize_nulls<ColumnDataType>(column, nulls);
+      _column_materializations[column_id] =
+          std::make_shared<ExpressionResult<ColumnDataType>>(std::move(values), std::move(nulls));
+
+    } else {
+      _column_materializations[column_id] = std::make_shared<ExpressionResult<ColumnDataType>>(std::move(values));
+    }
+  });
+}
+
+std::shared_ptr<ExpressionResult<std::string>> ExpressionEvaluator::_evaluate_substring(
+    const std::vector<std::shared_ptr<AbstractExpression>>& arguments) {
+  Assert(arguments.size() == 3, "SUBSTR expects three arguments");
+
+  const auto strings = evaluate_expression_to_result<std::string>(*arguments[0]);
+  const auto starts = evaluate_expression_to_result<int32_t>(*arguments[1]);
```

Comment:
`evaluate_expression_to_result` gives you a materialized result, correct? Isn't this overkill, especially if usually parameters 2 and 3 are literals?

Reply:
"evaluate_expression_to_result gives you a materialized result" nope, it doesn't, that's the trick. 
`evaluate_expression_to_result` gives you an `ExpressionResult`, yes, but both vectors of that ExpressionResult will have just one element: The literal value and whether its null. 

Still a bit overhead (2 vectors instead of 2 values), but in no way a fully "materialized result" as in one value per row, which I suppose is what you feared.